### DISCUSSION TITLE
Fast split logic global discussion

### DISCUSSION THREAD

**Discussion by Branikolog** [ihhub/fheroes2#2821] [Ideas]
Here it is! 
It's time to discuss the way we want to split creatures in different scenarios using keyboard and mouse.

**Comment by Branikolog** [comment]
First of all I want to thank our team and contributors, who help with finding and implementing the best solutions for this nice project!

**Comment by Branikolog** [comment]
I think, we should add always [2] button to any split dialog.
![image](https://user-images.githubusercontent.com/55348946/108491150-99814e00-72b4-11eb-80f2-433abe7286ac.png)
@LeHerosInconnu pushed a good idea: if we're splitting creature, change some values in dialog window, but changed our mind and want to split it into two separate equal stacks again (not 3, 4 or 5) we have to manually set the quantity of transferred creatures.
[on a screenshot we're trying to split 100 vampires. We manually set 15, but changed our mind and want to split them equally into two stacks. [2] button would help to recover the quantity without counting or re-making the dialog]
From my point I heavily insist on leaving the default separation of stack over empty slot in two equal sizes (So on default [2] button should always glare being selected), as this situation is the most common. I mean when we divide stack into two equal. In this case creature counter should display the quantity of creatures transferred equal to a half of original stack.

If we are going to split one stack of creatures over an other existing stack of creatures of the same type, within single army, the default transferring quantity should be 1 (minimal) as it shows how many creatures we want to add to target stack. (@LeHerosInconnu noted before, that this would help player to understand, that he is adding a certain amount of creatures to the target stack. My addition is to set the default not "0" (@LeHerosInconnu's proposal) but at least "1", so we can perform any action on pressing OK button from the default dialog state. If we want to transfer "0" creatures we can simply cancel the dialog. :D )
![Split one over another](https://user-images.githubusercontent.com/55348946/108492822-aacb5a00-72b6-11eb-81ec-3d2cd59df3e5.png)
In this case we should anyway have at least [2] button to be able to equalize both stacks and also (if possible) in 3, 4 or 5. In case of pressing 3, we should as a result have 3 equal stacks of creatures (counting the basic one, the target one and new one created in free slot. 
![Split one over another  3](https://user-images.githubusercontent.com/55348946/108493305-3644eb00-72b7-11eb-9589-188d515c88e9.png)
![Split one over another  3  result](https://user-images.githubusercontent.com/55348946/108493314-393fdb80-72b7-11eb-8e1e-2573e86fc12d.png)
This behaviour should be the same for splitting into 4 or 5 stacks. In the example above we should be able to divide creatures into 5 stack as well.

In a case of splitting one stack over an empty slot, while there's another stack of the same type in current army, we shouldn't consider that another stack and split the target stack into 2, 3 or 4 regarding overall free slots in current army. Which is the case currently.
If we want to create multiple equal stacks of the same type we first need to unite them. (except the case we're splitting over an other existing stack, which I reviewed above) (in a case we have only two stacks of the the same type)

Not to loose the functionality of this dialog window for the case, while transferring between two armies or between hero army and town garrison, I propose to leave all [2], [3] [etc.] buttons and allow player to split selected stack into several, leaving the first stack at the same place, and placing the rest into meeting hero/town garrison armies. The splitting into two stacks works currently as I mentioned before (Only [2] button is not implemented yet).
![Split one over empty slot of garrison](https://user-images.githubusercontent.com/55348946/108493916-fa5e5580-72b7-11eb-8e9a-a14c69bec0f7.png)
![Split one over empty slot of garrison  2](https://user-images.githubusercontent.com/55348946/108493930-fd594600-72b7-11eb-911d-57c3346c9af3.png)
![Split one over empty slot of garrison  2  result](https://user-images.githubusercontent.com/55348946/108493947-00543680-72b8-11eb-85c6-e80b67495aba.png)
In a case of splitting the basic stack into 3, 4, 5 or 6(new value) we should operate the same: select stack, select garrison army (or meeting hero) and press [3], [4], [5] or [6] button to make equal stacks leaving one in its basic position:
![Split one over empty slot of garrison](https://user-images.githubusercontent.com/55348946/108494401-896b6d80-72b8-11eb-8b05-79f19668ba90.png)
![image](https://user-images.githubusercontent.com/55348946/108494422-8cfef480-72b8-11eb-9f83-bd9b117a96fb.png)
![image](https://user-images.githubusercontent.com/55348946/108494465-9a1be380-72b8-11eb-86f7-d0eca5821436.png)

In the example above we should be able to split basic stack into 6 stacks: one + 5 in garrison.
![Split one over empty slot of garrison](https://user-images.githubusercontent.com/55348946/108495910-6c379e80-72ba-11eb-839a-4f6cd490edaf.png)
![image](https://user-images.githubusercontent.com/55348946/108495877-6346cd00-72ba-11eb-80ba-335c2188d8f4.png)

If we're going to transfer the basic stack in hero's army over an existing stack of the same type of garrison army we should also operate the same, as I've explained above: 
The default transferring quantity should be "1".
![image](https://user-images.githubusercontent.com/55348946/108496557-45c63300-72bb-11eb-891d-dcd4fe178fba.png)
While pressing [2], [3], [4] etc. fast split buttons we should create needed amount of equal stacks leaving one in our basic stack's position:
![image](https://user-images.githubusercontent.com/55348946/108496699-7c9c4900-72bb-11eb-8480-99733bc11e78.png)
![image](https://user-images.githubusercontent.com/55348946/108496737-8920a180-72bb-11eb-8c8a-4bc184a5de4e.png)
![image](https://user-images.githubusercontent.com/55348946/108496773-96d62700-72bb-11eb-87f1-2c9466aa63df.png)
[6] should also be available in this case.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

Nice post. :)
Everything seems good so far.
All that remains is to implement and test it. :)

**Reply by Branikolog** [reply]
@LeHerosInconnu 
Do you have any additions/objections or did I missed something?

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> @LeHerosInconnu
> Do you have any additions/objections or did I missed something?

No, if I understood everything correctly, it should work. :)

**Comment by Branikolog** [comment]
Also, I remember a brilliant suggestion of @LeHerosInconnu regarding changing of display quantity in creature counter, while pressing [2], [3], [4] or [5] fast split buttons, which would show the result stack size while splitting into several equal. Do we have an opened issue for that?

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Also, I remember a brilliant suggestion of @LeHerosInconnu regarding changing of display quantity in creature counter, while pressing [2], [3], [4] or [5] fast split buttons, which would show the result stack size while splitting into several equal. Do we have an opened issue for that?

Thanks, :) this one: https://github.com/ihhub/fheroes2/issues/2406.

**Reply by Branikolog** [reply]
Oh, nice!
I have a small improvement for this issue:
In a case of pressing any fast split button I propose to change the title of the window "How many troops to move?" so it corresponds the action we're choosing: "Create X groups of troops". (As usual, I ask for help with text description, because of my poor english)
So below the new description we would see the size of each stack inside of a troop counter, which should be updated, while choosing different fast split buttons.
In a case we start pressing arrows to change the amount of transferred troops, or using scroll wheel we should cancel fast split button selection and return original title in the upper part of window back.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Oh, nice!
> I have a small improvement for this issue:
> In a case of pressing any fast split button I propose to change the title of the window "How many troops to move?" so it corresponds the action we're choosing: "Create X groups of troops". (As usual, I ask for help with text description, because of my poor english)
> So below the new description we would see the size of each stack inside of a troop counter, which should be updated, while choosing different fast split buttons.
> In a case we start pressing arrows to change the amount of transferred troops, or using scroll wheel we should cancel fast split button selection and return original title in the upper part of window back.

Good idea too. :)

Something like this:

"Divide the troop into X troops."
"Divide the troop into X groups."

"Separate the troop into X troops."

"Split the troop into X troops."

"Distribute the troop into X troops."

The word "groups" can be used to avoid repeating the word "troops", but using the word "troops" twice doesn't shock me.

I would prefer:

"Split the troop into X troops."

Or:

"Divide the troop into X troops."

**Reply by Branikolog** [reply]
Sounds good.
Is it possible to add phrase to a sentence, where the last word inteded to be the number of creatures in each stack? So the counter below will be the part of the whole sentence.
Something like: " ... with a stack size of: "
And it should be short enough to fit into single line...
Could you, please, advise something?
If we don't have anough place for that, I like all the variations above.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Sounds good.
> Is it possible to add phrase to a sentence, where the last word inteded to be the number of creatures in each stack? So the counter below will be the part of the whole sentence.
> Something like: " ... with a stack size of: "
> And it should be short enough to fit into single line...
> Could you, please, advise something?
> If we don't have anough place for that, I like all the variations above.

This seems difficult without changing the width of the window. :D

The sentence "Split the troop into X troops." may already be too long.

But the reduced sentence "Split into X troops of:" should be short enough. :)

**Reply by Branikolog** [reply]
> "Split into X troops of:"

That looks quite short and informative for me!
